feat: Implement LaTeX generation

hiyaryan opened the pull request:
This draft PR transfers the pyQuil LaTeX generation feature to quil-rs.

The initial commit,
1. Adds a new `latex.rs` to the `program` crate.
2. Exposes it as a feature to the project in `Cargo.toml` and `mod.rs`.
3. Includes an initial snapshot test for the simple Program `CONTROLLED H 3 2`.
---
program/latex.rs
This file has two sections with some initial code, 
1. LaTeX Implementation 
2. Testing Suite 

LaTeX Implementation
The LaTeX implementation section of `latex.rs` is where the LaTeX generation will occur. The current template was used to generate the first snapshot test. It includes DiagramSettings, directly taken from [pyQuil](https://github.com/rigetti/pyquil/blob/master/pyquil/latex/_diagram.py), that will allow users to customize the LaTeX for their Program following the [Builder](https://rust-unofficial.github.io/patterns/patterns/creational/builder.html) design pattern in Rust. The template also contains an outline for displaying various `Result::Err` using [thiserror](https://docs.rs/thiserror/latest/thiserror/) on a `LatexGenError` enum. 

Testing Suite
The first test in the testing module containing the suite of LaTeX unit tests is a simple snapshot test for `CONTROLLED H 3 2`. The snap file for this test is program/snapshots/quil_rs__program__latex__tests__controlled_gate.snap. It was generated by passing the following expected LaTeX to the unit test `test_controlled_gate` as the initial snapshot assertion.

```latex
\[convert={density=300,outext=.png}]{standalone}
\usepackage[margin=1in]{geometry}
\usepackage{tikz}
\usetikzlibrary{quantikz}
\begin{document}
\begin{tikzcd}
\lstick{\ket{q_{2}}} & \gate{H} & \qw \\
\lstick{\ket{q_{3}}} & \ctrl{-1} & \qw
\end{tikzcd}
\end{document}
```
---
Main Tasks and Sub Tasks lists will be used to help keep track of the work. Once all tasks are marked as complete in the Main Tasks the stage of this PR can be changed from draft to ready for review.

Main Tasks
- [ ] I. Implement all LaTeX generation features currently available in pyQuil.
- [ ] II. Implement new CCNOT and CPHASE features.


Sub Tasks (live list)

I. Implement all LaTeX generation features currently available in pyQuil.
- [ ] Research external LaTeX generation crates.
- [ ] Create simple snapshot tests for single qubit Programs.
- [ ] Implement LaTeX generation for single qubit Programs.
...
- [ ] Pass assertion for multi-qubit `test_controlled_gate` test.
...

II. Implement new CCNOT and CPHASE features
...

hiyaryan commented:
Having looked through the available, highest All-Time LaTeX crates and keeping in mind when it was last updated, the following stood out with the second being my choice (happy to discuss why it shouldn't be).

LaTeX Crates
- [tectonic](https://crates.io/crates/tectonic) (ideal choice)
Tectonic definitely is the most supported, well-documented, and expansive TeX/LaTeX tool (actually, engine). However, I think it might be offering more than is needed for our use case based on the number of external C/C++ libraries it relies on (see: [building tectonic](https://tectonic-typesetting.github.io/book/latest/howto/build-tectonic/index.html)). While it might be nice to have all of these LaTeX features accessible through a CLI that quil-rs can deliver at some point, adding it to the project with all of its own dependencies is not so straightforward. Really it seems to advertise itself as a [standalone CLI application](https://tectonic-typesetting.github.io/en-US/).

- [latex](https://crates.io/crates/latex) (my choice)
This latex crate works right out of the box, no separate `cargo install` needed. Had practically no issues adding it as an optional dependency to the Cargo.toml file. Was also able to generate some LaTeX almost immediately compared to the almost 2-hours trying to get tectonic to work. It's a very lightweight LaTeX tool that essentially builds your LaTeX for you, which is what we want. The [documentation](https://docs.rs/latex/0.3.1/latex/) is not expansive and the last update was made to it over 3 years ago (not great), however, that's reasonable for what it does and how simple it is to use. With what it looks to offer, I think it can do everything we need it to that pyQuil LaTeX generation does. There's some nice [examples](https://github.com/Michael-F-Bryan/latex-rs/tree/master/examples) too.

hiyaryan commented:
While trying to use the latex crate added in the second to last commit, there doesn't seem to be a straightforward way to pass parameters to a particular LaTeX object. For example when trying to write the diagram below for a single qubit passing through a Y gate, the process is to create a new Document and instantiate it with a `DocumentClass`. This can be classified as standalone using the `DocumentClass::Other("standalone".to_string())`. It generates `\documentclass{standalone}`. In order to produce the same diagrams generated in pyQuil, these parameters, `[convert={density=300,outext=.png}]` must be included in between the document object and the type. When looking into the crates documentation, it doesn't seem like there is a straightforward way to include parameters in square brackets. None of the examples provided include parameters neither. I have tried reaching out to the developer to see if they can provide some direction. If this is not possible, the next alternative is to create my own LaTeX document builder similar to how pyQuil generates the LaTeX. This might be simpler and more ideal than attempting to use, conform to, and learn an external LaTeX generation tool.

```latex
\documentclass[convert={density=300,outext=.png}]{standalone}
\usepackage[margin=1in]{geometry}
\usepackage{tikz}
\usetikzlibrary{quantikz}
\begin{document}
\begin{tikzcd}
\lstick{\ket{q_{1}}} & \gate{Y} & \qw
\end{tikzcd}
\end{document}
```

kalzoo commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,173 @@
+        use crate::program::latex::TikzOperator;
+
+        #[test]
+        fn test_tikz_left_ket() {
```
Worth noting before you write too many more of these - macros are quite useful for test case patterns like this. There is a learning curve but it's a powerful tool in the kit

hiyaryan replied:
Thanks for watching out Kalan! I haven't written any more tests since I started the new design. I'll look into macros once I get around to writing more.

hiyaryan commented:
The previous commit is the new model I designed for the framework that the LaTeX generation can be built upon. Everything functions exactly how it previously did before this commit with some refactoring to match the design pattern discussed in this [README](https://github.com/hiyaryan/quil-rs/tree/to-latex/src/program/latex). Please note that it is still very much a work in progress. This initial commit is only meant to highlight the direction that I am moving in. Specific details need to be more fully fleshed out, more comments need to be added and the README needs to be polished, and the default to_latex functionality that returns the Quantikz diagram pyQuil currently produces needs to be implemented as a concrete product with the framework in place for it. Below is a diagram that shows the general structure of Quil LaTeX feature. It is not exactly how you might use UML, neither can it be really considering Rust is not exactly a OOP language. Note that it is not all inclusive, so some of the functions and attributes are not labeled (this will be updated).

![quil-rs LaTeX drawio](https://user-images.githubusercontent.com/58452495/219515601-388329ac-65b9-4d9e-b711-ed68b9340a66.png)

Shadow53 commented on src/program/latex.rs:
```diff
@@ -0,0 +1,118 @@
+use crate::Program;
+use generation::{LatexFactory, diagram::DiagramFactory};
```
FWIW, Rust uses the convention of `XBuilder` instead of `XFactory`

Shadow53 commented on src/program/latex.rs:
```diff
@@ -0,0 +1,118 @@
+pub trait Latex {
+    /// Returns a Result containing a string of LaTeX or a LatexGenerationError.
+    ///
+    /// # Arguments
```
Generally don't need to describe the arguments in doc comments, as they are visible in the generated documentation.

Also, this comment seems outdated anyways.

Shadow53 commented on src/program/latex/generation.rs:
```diff
@@ -0,0 +1,7 @@
+pub mod diagram;
+
+pub trait LatexFactory {
+    fn create_tikz_diagram(&self) -> Box<dyn Tikz>;
```
You probably want to use associated types instead of boxed trait objects:

```suggestion
    type Tikz: Tikz;
    fn create_tikz_diagram(&self) -> Self::Tikz;
```

Shadow53 commented on src/program/latex/generation/diagram/tikz.rs:
```diff
@@ -0,0 +1,80 @@
+impl TikzOperator {
+    fn get_tikz_operator(tikz_operator: Self) -> String {
+        match tikz_operator {
+            Self::TikzLeftKet(qubit) => format(format_args!(r#"\lstick{{\ket{{q_{{{qubit}}}}}}}"#)), // \lstick{\ket{q_{qubit}}}
```
```suggestion
            Self::TikzLeftKet(qubit) => format!(r#"\lstick{{\ket{{q_{{{qubit}}}}}}}"#), // \lstick{\ket{q_{qubit}}}
```

Shadow53 replied:
Same goes for the below

Shadow53 commented on src/program/latex.rs:
```diff
@@ -0,0 +1,118 @@
+
+pub enum Document {
+    Diagram,
+    None,
```
Elsewhere seems to indicate that `None` causes errors -- is there a case where it is valid? If not, why include it?

Shadow53 commented on src/program/latex/generation/diagram/tikz.rs:
```diff
@@ -0,0 +1,80 @@
+    TikzMeasure,
+}
+
+impl TikzOperator {
```
This might be better off as an `impl fmt::Display for TikzOperator` block, which gives you an implementation of `ToString` for free

Shadow53 reviewed (commented):
Left some comments on how to make this more Rusty.

I'd also like to suggest looking into the builder pattern in Rust, which is (1) more conventional than a Java-like Factory setup and (2) may help avoid errors from bad user input by just not allowing it.

hiyaryan commented:
Quick summary of all commits over the past two weeks.

- Reverted previous model using the abstract factory pattern to a simpler model matching, more closely, that of pyQuil's LaTeX generation model. This ensures that there is no dead code waiting for features to be implemented that may ultimately not be.
- More closely modeled this version of to_latex to the Quantikz [documentation](https://arxiv.org/pdf/1809.03842.pdf). Changed what pyQuil called TIKZ_<Operators> to the official `Command` name defined in the docs linked. This helps map the latex generation code to Quantikz which is based on the Quantikz LaTeX package making it easier to implement. 
- Developed the data structures for a `Diagram` that contains a circuit which is a `Wire` collection. Currently, the `Diagram` has two purposes: format the circuit to a String of LaTeX, and hold information about the relationships between the wires. The purpose of the `Wire` is to hold information about all items (e.g. gates, measures, do nothings) on a *single* wire.
- Currently, `Diagram` can take any number of quil instructions of the form `<Gate> <Qubit>` and return the corresponding LaTeX String as of the last two commits. This does not require `Diagram` to hold any information about the relationships between the wires since in these instructions there are none.

Future commits might include more tests specific to the functions specified in this summary or be encapsulated in tests related to the next step in the LaTeX generation feature which is creating more complex LaTeX diagrams showing relationships between multiple qubits.

hiyaryan commented:
Here is my solution to the order preservation issue stated in the last comment. Use a `Vec<String>`, called `order`, to index all of the keys of the HashMap. Since all of the keys in the HashMap are iterated over anyway in order to format each `Wire` in `Diagram.circuit` into LaTeX, a `Vec<String>` is essentially a copy of the key list whose index is used to preserve the order at which the wire was added to the circuit. So iterating over `Vec<String>` with `for key in &diagram.order` provides the same logic in accessing items in the HashMap taking advantage of O(1) search, and is actually faster, O(len), iterating over as `keys`—which has a time complexity of O(capacity) [ref](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.keys:~:text=iter_mut-,keys,-len). While this solution does improve the performance of the algorithm, it worsens the space complexity with the additional attribute, `order: Vec<String>`, added to `Diagram`.

hiyaryan commented:
CNOT LaTeX generation now works for variable qubits. The LaTeX generated is as follows,

```latex
\lstick{\ket{q_{0}}} & \ctrl{1} & \qw \\
\lstick{\ket{q_{1}}} & \targ{} & \qw
```

```latex
\lstick{\ket{q_{1}}} & \ctrl{1} & \qw \\
\lstick{\ket{q_{0}}} & \targ{} & \qw
```

If we want to preserve the order of the qubits, fixed qubits, the last latex above should be generated as,

```latex
\lstick{\ket{q_{0}}} & \ctrl{0} & \qw \\
\lstick{\ket{q_{1}}} & \targ{-1} & \qw
```

All this is doing is looking at the previous line indicated by the -1 which Quantikz represents as a solid line drawn from a solid dot from ket(1) to an open dot on ket(0). Preserving this order for fixed qubits is in work from this point and will be added as a default setting.

hiyaryan commented:
The last 6 commits implements `CNOT` and includes the the `impute_missing_qubits` setting that inserts qubits between e.g. `CNOT 2 5`. With this setting this makes it so that the Diagram renders wires labeled 2 , 3, 4, 5 instead of 2 and 5 without the setting. This setting allowed for thorough testing (see added TDD snapshot tests) of `CNOT` since every time a new wire is added the target had to be updated. The target qubit still remains the same, but the wire it lies on is now further from the control. Quantikz uses the distance between stacked wires with a negative value indicated a target above the control qubit, and a positive value below the control qubit. The algorithm that sets the relationship between control and target qubits is designed to be independent of whether `imput_missing_qubits` is set to `true` or `false`.

As an additional note, the `Diagram.circuit` was updated from a `HashMap<u64, Wire>` to a `BTreeMap<u64, Wire>` which automatically sorts the wires upon insertion by qubit number `u64`. The `Diagram.order` is still required though. `Diagram.order` maintains the insertion order of the instruction qubits. For example, if an instruction with a set of qubits in this order [5, 2] is inserted into BTreeMap, the map sorts it to [2, 5]. This is helpful when rendering the circuits in increasing order, however, the order of qubits in an instruction is important because `CNOT 5 2` is different from `CNOT 2 5`. This is the insertion order that `Diagram.order` is preserving. This ordering comes from the fact that the Fixed qubits from e.g. `CNOT 5 2` are given in the order [5, 2]. This is the `order` that needs to be preserved in order for the Diagram to know which qubit is the control and which is the target when parsing the wires prior to formatting the circuit into LaTeX string in increasing order which is the order given by the `BTreeMap`.

hiyaryan commented:
Summary of the last 27 commits

I. First 15 commits implement 
`Phase`, Controlled Phase (`CZ`, `CPHASE`) and Controlled X (`CNOT`, `CCNOT`), modifiers (`CONTROLLED`, `DAGGER`), and Setting `texify_numerical_constants`. 

II. Next 7 commits
Debug all the previous commits with TDD using Quil programs taken from [good-test-files](https://github.com/quil-lang/quilc/tree/a2925c5dd1764534ef8ea4e66b051cc43cecea30/tests/good-test-files). The biggest change occurred in this commit, [Refactor to_latex with initial circuit of used qubits.](https://github.com/rigetti/quil-rs/pull/145/commits/2338b2e45d026ef69d49b1b87efd624b04fce2d7) changing the `to_latex` algorithm which takes account of all qubits at the beginning of the program as opposed to adding each qubit to the circuit one at a time per instruction.

III. Last 5 commits 
Include all of the *good tests* of programs that can be generated in this implementation of to LaTeX which pass as a result of the refactoring from the previous 7 commits. These 5 commits also include error handling which decidedly doesn't allow Programs containing unsupported gates or modifiers to *try* to be generated (while any program can be passed through this feature, these unsupported Programs will not generate the correct circuit). Finally, the documentation is updated which elaborates more on various functions and supported Programs and cleans up various parts of the code.

hiyaryan commented:
Here is a list of Programs with their associated LaTeX and rendered images.

1. Standard Gate Program
```
H 0
CNOT 0 1
X 1
CNOT 1 2
```
```latex
\documentclass[convert={density=300,outext=.png}]{standalone}
\usepackage[margin=1in]{geometry}
\usepackage{tikz}
\usetikzlibrary{quantikz}
\begin{document}
\begin{tikzcd}
\lstick{\ket{q_{0}}} & \gate{H} & \ctrl{1} & \qw & \qw & \qw \\
\lstick{\ket{q_{1}}} & \qw & \targ{} & \gate{X} & \ctrl{1} & \qw \\
\lstick{\ket{q_{2}}} & \qw & \qw & \qw & \targ{} & \qw
\end{tikzcd}
\end{document}
```
<img width="564" alt="Screenshot 2023-03-15 at 6 42 17 PM" src="https://user-images.githubusercontent.com/58452495/225488405-7018ef07-f2e7-4a0a-be6e-e272eb89fe28.png">


2. Bell State with Imputed Qubits
```
H 5
CNOT 5 2
```
```latex
\documentclass[convert={density=300,outext=.png}]{standalone}
\usepackage[margin=1in]{geometry}
\usepackage{tikz}
\usetikzlibrary{quantikz}
\begin{document}
\begin{tikzcd}
\lstick{\ket{q_{2}}} & \qw & \targ{} & \qw \\
\lstick{\ket{q_{3}}} & \qw & \qw & \qw \\
\lstick{\ket{q_{4}}} & \qw & \qw & \qw \\
\lstick{\ket{q_{5}}} & \gate{H} & \ctrl{-3} & \qw
\end{tikzcd}
\end{document}
```
<img width="555" alt="Screenshot 2023-03-15 at 6 42 46 PM" src="https://user-images.githubusercontent.com/58452495/225488458-e6a14e05-231e-474f-b0d7-fc29a43d0586.png">


3. Good modifiers
```
Y 0
CONTROLLED Y 0 1
CONTROLLED CONTROLLED Y 0 1 2
CONTROLLED CONTROLLED CONTROLLED Y 0 1 2 3
DAGGER Y 0
DAGGER DAGGER Y 0
DAGGER DAGGER DAGGER Y 0
CONTROLLED DAGGER Y 0 1
CONTROLLED DAGGER CONTROLLED Y 0 1 2
CONTROLLED DAGGER CONTROLLED DAGGER Y 0 1 2
DEFGATE G AS MATRIX:
        1,0
        0,1

CONTROLLED G 0 1
DAGGER G 0
```
```latex
\documentclass[convert={density=300,outext=.png}]{standalone}
\usepackage[margin=1in]{geometry}
\usepackage{tikz}
\usetikzlibrary{quantikz}
\begin{document}
\begin{tikzcd}
\lstick{\ket{q_{0}}} & \gate{Y} & \ctrl{1} & \ctrl{2} & \ctrl{3} & \gate{Y^{\dagger}} & \gate{Y^{\dagger}^{\dagger}} & \gate{Y^{\dagger}^{\dagger}^{\dagger}} & \ctrl{1} & \ctrl{2} & \ctrl{2} & \ctrl{1} & \gate{G^{\dagger}} & \qw \\
\lstick{\ket{q_{1}}} & \qw & \gate{Y} & \ctrl{1} & \ctrl{2} & \qw & \qw & \qw & \gate{Y^{\dagger}} & \ctrl{1} & \ctrl{1} & \gate{G} & \qw & \qw \\
\lstick{\ket{q_{2}}} & \qw & \qw & \gate{Y} & \ctrl{1} & \qw & \qw & \qw & \qw & \gate{Y^{\dagger}} & \gate{Y^{\dagger}^{\dagger}} & \qw & \qw & \qw \\
\lstick{\ket{q_{3}}} & \qw & \qw & \qw & \gate{Y} & \qw & \qw & \qw & \qw & \qw & \qw & \qw & \qw & \qw
\end{tikzcd}
\end{document}
```
<img width="563" alt="Screenshot 2023-03-15 at 6 43 07 PM" src="https://user-images.githubusercontent.com/58452495/225488486-0b4ce728-70ef-46d1-bc3c-cbc5de1af7a8.png">


4. DEFGATE
```
DEFGATE H0 AS MATRIX:
        0.707,0.707
        0.707,(-0.707)

DEFGATE H1 AS MATRIX:
        (1/sqrt(2)),(1/sqrt(2))
        (1/sqrt(2)),((-1)/sqrt(2))

H0 0
H1 1
```
```latex
\documentclass[convert={density=300,outext=.png}]{standalone}
\usepackage[margin=1in]{geometry}
\usepackage{tikz}
\usetikzlibrary{quantikz}
\begin{document}
\begin{tikzcd}
\lstick{\ket{q_{0}}} & \gate{H0} & \qw & \qw \\
\lstick{\ket{q_{1}}} & \qw & \gate{H1} & \qw
\end{tikzcd}
\end{document}
```
<img width="564" alt="Screenshot 2023-03-15 at 6 43 31 PM" src="https://user-images.githubusercontent.com/58452495/225488509-d85ceb51-bc17-4306-9884-75f1817f7d59.png">


5. CPHASE with `pi` Rotation
```
CPHASE(pi) 0 1
```
```latex
\documentclass[convert={density=300,outext=.png}]{standalone}
\usepackage[margin=1in]{geometry}
\usepackage{tikz}
\usetikzlibrary{quantikz}
\begin{document}
\begin{tikzcd}
\lstick{\ket{q_{0}}} & \ctrl{1} & \qw \\
\lstick{\ket{q_{1}}} & \phase{\pi} & \qw
\end{tikzcd}
\end{document}
```
<img width="560" alt="Screenshot 2023-03-15 at 6 43 51 PM" src="https://user-images.githubusercontent.com/58452495/225488530-0e4e88e2-1f28-49e5-93b4-177dbc1fd8f8.png">


6.  Toffoli Gate
```
CCNOT 1 2 0
```
```latex
\documentclass[convert={density=300,outext=.png}]{standalone}
\usepackage[margin=1in]{geometry}
\usepackage{tikz}
\usetikzlibrary{quantikz}
\begin{document}
\begin{tikzcd}
\lstick{\ket{q_{0}}} & \targ{} & \qw \\
\lstick{\ket{q_{1}}} & \ctrl{-1} & \qw \\
\lstick{\ket{q_{2}}} & \ctrl{-2} & \qw
\end{tikzcd}
\end{document}
```
<img width="559" alt="Screenshot 2023-03-15 at 6 44 11 PM" src="https://user-images.githubusercontent.com/58452495/225488569-b4d4792f-99e4-48e5-93f0-dbe21a2499d6.png">


7. CPHASE with `alpha` Rotation Texify: Off (if Texify: On, will be rendered with Greek symbol `α`)
```
CPHASE(alpha[0]) 0 1
```
```latex
\documentclass[convert={density=300,outext=.png}]{standalone}
\usepackage[margin=1in]{geometry}
\usepackage{tikz}
\usetikzlibrary{quantikz}
\begin{document}
\begin{tikzcd}
\lstick{\ket{q_{0}}} & \ctrl{1} & \qw \\
\lstick{\ket{q_{1}}} & \phase{\text{alpha}} & \qw
\end{tikzcd}
\end{document}
```
<img width="561" alt="Screenshot 2023-03-15 at 6 44 36 PM" src="https://user-images.githubusercontent.com/58452495/225488592-cc4d7166-cc54-4a4f-afde-49b301812b45.png">


8. Label Qubit Lines Off
```
H 0
CNOT 0 1
```
```latex
\documentclass[convert={density=300,outext=.png}]{standalone}
\usepackage[margin=1in]{geometry}
\usepackage{tikz}
\usetikzlibrary{quantikz}
\begin{document}
\begin{tikzcd}
\qw & \gate{H} & \ctrl{1} & \qw \\
\qw & \qw & \targ{} & \qw
\end{tikzcd}
\end{document}
```
<img width="563" alt="Screenshot 2023-03-15 at 6 44 54 PM" src="https://user-images.githubusercontent.com/58452495/225488616-8816d056-2119-4721-b220-1d2172348944.png">

notmgsk commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1600 @@
+    ///
+    /// # Arguments
+    /// `text` - a String representing a greek or alaphanumeric symbol
+    pub fn match_symbol(text: String) -> Symbol {
```
This could be replaced with a `impl From<String> for Symbol`, or you could probably use the `strum` package to generate that `impl` for you (if you don't mind adding an extra dependency).

notmgsk replied:
But then you need an error, even though there is no error case.

notmgsk commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1600 @@
+    /// # Arguments
+    /// `text` - a String representing a greek or alaphanumeric symbol
+    pub fn match_symbol(text: String) -> Symbol {
+        if text == "alpha" {
```
This is more typically expressed as a `match`.

notmgsk replied:
@hiyaryan you resolved the conversation, but it still appears to be an if/else chain rather than a `match`.

notmgsk replied:
Maybe you didn't push the changes

notmgsk commented on src/program/latex/mod.rs:
````diff
@@ -0,0 +1,1600 @@
+    ///     &Parameter::Symbol(Symbol::Alpha)
+    /// );
+    /// ```
+    pub fn get_symbol(symbol: &Parameter) -> String {
````
This is confusing to me. There are three types involved: `&Parameter`, `String`, and `Symbol`. The function is named `get_symbol` but returns `String`.

I think this would be tidier as a `impl ToString` on _both_ `Parameter` and `Symbol`. (The `Parameter` impl would call the `Symbol` impl.)

hiyaryan replied:
Great solution! That really resolved some of the confusion.

notmgsk replied:
@Shadow53 I gotta remember that tip. Ty

notmgsk commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1600 @@
+/// Settings contains the metadata that allows the user to customize how the
+/// circuit is rendered or use the default implementation.
+#[derive(Debug)]
+pub struct Settings {
```
`RenderSettings`

notmgsk commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1600 @@
+    ///
+    ///  # Arguments
+    /// `name` - name of the qubit
+    pub fn label_qubit_lines(&self, name: u64) -> String {
```
This seems unrelated to the intent of `Settings`

hiyaryan replied:
My intent with `Settings` is that if the user decides to use other than `default`, the Settings impl block will contain the functionality to apply those changes. It's just a way to group these effects into a single cohesive location.

notmgsk commented on src/program/latex/mod.rs:
````diff
@@ -0,0 +1,1600 @@
+    /// };
+    /// program.to_latex(settings).expect("");
+    /// ```
+    pub fn impute_missing_qubits(&self, column: u32, circuit: &mut BTreeMap<u64, Box<Wire>>) {
````
Likewise, not sure this needs to be defined in this `impl`. I think it could just be a top-level function with no `&self` parameter (doesn't look like that is used anywhere).

hiyaryan replied:
(Addition to previous comment in this block) You're right, `&self` should be removed. These are just utility functions that  `Settings` can apply. Does it make more sense to take it out of the Settings block?

notmgsk commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1600 @@
+///     Single wire commands: lstick, rstick, qw, meter
+///     Multi-wire commands: ctrl, targ, control, (swap, targx)
+#[derive(Debug)]
+pub enum Command {
```
Does this need to be `pub`?

As a consumer, the most I would expect to be public from this module is the entry point to convert a `quil_rs::Program` to a (LaTeX) `String` (and thus the required `Settings`). All other types should be private or `pub(crate)`.

notmgsk reviewed (changes requested):
Partial review! Will do more later today.

MarquessV commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1600 @@
+    fn set_qw(&mut self, qubits: &HashSet<Qubit>, instruction: &Instruction) {
+        for program_qubit in qubits {
+            let mut found = false;
+            match &instruction {
```
Two things:

(1) If this function takes a single `Instruction`, and that `Instruction` has to be a `Gate`, then why bother checking for every qubit in `qubits`? Would it make more sense to do the match first, and have the loop in the body?

(2) The `HashSet` API has a [`contains(other)`](https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.contains) method. Why not use that instead of managing the found boolean and nesting loops? Something like:

```rs
match &instruction {
    instruction::Instruction::Gate(gate) => {
        for gate_qubit in &gate.qubits {
            if !qubits.contains(gate_qubit) {
                match program_qubit {
                    Instruction::Qubit::Fixed(q) => {
                        self.circuit
                            .get_mut(&q)
                            .and_then(|wire| wire.empty.insert(self.column, Command::Qw));
                    },
                    _ => (),
                }
            }
        }
    },
    _ => (),
}

MarquessV commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1600 @@
+    /// otherwise, it is negative. See [`Quantikz`] documentation on CNOT for
+    /// some background that helps justify this approach.
+    ///
+    /// This function is expensive with a time complexity of O(n^2). In the
```
We may be able to avoid this costly function by determining the size of the circuit up front. If we know how many wires/rows the circuit will have, then the `ctrl` and `targ` values can be calculated as the difference between the qubit indices for the gate.

MarquessV reviewed (commented):
Awesome job getting all those tests passing! Left a partial review as well. Will circle back later.

Shadow53 commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1600 @@
+    /// `qubits` - qubits used in the quil program
+    /// `instruction` - exposes qubits in a single instruction
+    fn set_qw(&mut self, qubits: &HashSet<Qubit>, instruction: &Instruction) {
+        for program_qubit in qubits {
+            let mut found = false;
+            match &instruction {
+                instruction::Instruction::Gate(gate) => {
+                    for gate_qubit in &gate.qubits {
+                        if program_qubit == gate_qubit {
+                            found = true;
+                        }
+                    }
+
+                    if !found {
+                        match program_qubit {
+                            instruction::Qubit::Fixed(q) => {
+                                self.circuit
+                                    .get_mut(&q)
+                                    .and_then(|wire| wire.empty.insert(self.column, Command::Qw));
+                            }
+                            _ => (),
+                        }
+                    }
+                }
+                _ => (),
+            }
+        }
```
A few things here:

- If matching a single variant, prefer `if let` instead.
- Instead of using a boolean flag, use loop labels to `continue` the outer loop
- Style preference, `and_then` is more for `Option` combinatorics where you care about the returned value. In the case of this code, I think `if let Some(_)` is more appropriate

```suggestion
        'program_loop: for program_qubit in qubits {
            if let Instruction::Gate(gate) = instruction {
                for gate_qubit in &gate.qubits {
                    if program_qubit == gate_qubit {
                        continue 'program_loop;
                    }
                }

                if let Qubit::Fixed(q) = program_qubit {
                    if let Some(wire) = self.circuit.get_mut(&q) {
                        wire.empty.insert(self.column, Command::Qw);
                    }
                }
            }
        }
```

Shadow53 commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1600 @@
+    /// instead each of the modifiers are added to the wire at the current
+    /// column where it is to be applied using the Command::Super variant.
+    ///
+    /// # Arguments
+    /// `&mut self` - exposes the current column of the Circuit
+    /// `gate` - exposes the modifiers associated with the instruction gate
+    /// `wire` - exposes the wire to be pushed to in Circuit
```
Again, don't think this is necessary

hiyaryan replied:
Removed the middle paragraph of the docs [Use write! in Diagram fmt, reduce/reformat docs.](https://github.com/rigetti/quil-rs/pull/145/commits/6a2fd9e508c61782156ac7e9e3829320153c0df8)

Shadow53 commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1600 @@
+            for modifer in &gate.modifiers {
+                match modifer {
+                    instruction::GateModifier::Dagger => {
+                        if let Some(modifiers) = wire.modifiers.get_mut(&self.column) {
+                            modifiers.push("dagger".to_string())
+                        } else {
+                            wire.modifiers
+                                .insert(self.column, vec!["dagger".to_string()]);
+                        }
```
Behold, the `entry` API!

This is not necessarily _better_ (there's a potentially unnecessary `clone`), but it's a good API to be familiar with.

```suggestion
                        wire.modifiers.entry(self.column.clone())
                            .and_modify(|modifiers| modifiers.push("dagger".to_string()))
                            .or_insert_with(|| vec!["dagger".to_string()]);
```

Shadow53 commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1600 @@
+                        // get each existing qubit in the circuit
+                        if *wire.0 == ctrl || *wire.0 == targ {
+                            // if the qubit is the ctrl or target
+                            if let Some(_) = open {
```
```suggestion
                            if open.is_some() {
```

Shadow53 commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1600 @@
+
+                    // find the range between the qubits
+                    let mut i = 0;
+                    for wire in &self.circuit {
+                        // get each existing qubit in the circuit
+                        if *wire.0 == ctrl || *wire.0 == targ {
+                            // if the qubit is the ctrl or target
+                            if let Some(_) = open {
+                                close = Some(i);
+                                break;
+
+                            // open qubit in range not found, set open qubit
+                            } else {
+                                open = Some(i)
+                            }
+                        }
+
+                        i += 1;
```
Two things:

- `Iterator::enumerate()` will return an `i` for you so you don't need to calculate it yourself
- Prefer `Option::is_some()` if you don't care about the contained value

```suggestion
                    for (i, wire) in self.circuit.iter().enumerate() {
                        // get each existing qubit in the circuit
                        if *wire.0 == ctrl || *wire.0 == targ {
                            // if the qubit is the ctrl or target
                            if open.is_some() {
                                close = Some(i);
                                break;

                            // open qubit in range not found, set open qubit
                            } else {
                                open = Some(i)
                            }
                        }
```

Shadow53 commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1530 @@
+                    }
+                    // set wire at column as the control qubit of target qubit
+                    // computed as the distance from the control qubit
+                    self.circuit
+                        .get_mut(&ctrl)
+                        .and_then(|wire| wire.ctrl.insert(c, vector));
```
Same style suggestion as above to use `if let`

Shadow53 commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1530 @@
+        let qubit = wire.name;
+
+        // find wire in circuit collection
+        match self.circuit.get_mut(&wire.name) {
+            Some(wire_in_circuit) => {
```
`if let Some(wire_in_circuit)`

Shadow53 commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1530 @@
+            // get the newly added gate if any at the column it was added
+            if let Some(gate) = wire.gates.get(&self.column) {
+                // tag relationships for multi qubit gates
+                if gate.starts_with('C') {
```
This seems suspect. What happens if someone `DEFGATE`s a new gate with a name starting with `C`? Would be better to have an `is_controlled()` method or something to check this.

Shadow53 commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1530 @@
+impl Display for Diagram {
+    /// Returns a result containing the Diagram Circuit as LaTeX string which
+    /// can be input into the body of the Document.
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
```
Rather than build a `String` and write it at the end, just `write!()` the parts you'd be adding to the string. It is more straightforward and less expensive.

hiyaryan replied:
Refactored using `write!()` as opposed to building a `String` [Use write! in Diagram fmt, reduce/reformat docs.](https://github.com/rigetti/quil-rs/pull/145/commits/6a2fd9e508c61782156ac7e9e3829320153c0df8)

Shadow53 commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1530 @@
+    /// can be input into the body of the Document.
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // add a newline between the first line and the header
+        let mut body = String::from('\n');
```
For example:

```suggestion
        writeln!(f)?;
```

Shadow53 commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1530 @@
+                            }
+                        }
+
+                        if gate.starts_with('C') {
```
Don't we have a `Gate` type? Why aren't we using that instead of strings?

Shadow53 commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1530 @@
+    empty: HashMap<u32, Command>,
+}
+
+impl Default for Wire {
+    fn default() -> Self {
+        Self {
+            name: 0,
+            gates: HashMap::new(),
+            ctrl: HashMap::new(),
+            targ: HashMap::new(),
+            parameters: HashMap::new(),
+            modifiers: HashMap::new(),
+            empty: HashMap::new(),
+        }
+    }
+}
```
Just use `derive(Default)` instead

Shadow53 commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1530 @@
+    ControlledX(String),
+    DefGate(String),
+    Modifiers(String),
+    Unsupported(String), // for error handling
```
If this has an `Unsupported` variant, the `enum` should not be called `SupportedGate`

Shadow53 commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1530 @@
+    ///
+    /// # Arguments
+    /// `name` - the name of the gate from instruction
+    fn get_supported_standard_gate(name: String) -> Self {
+        if name == "I" || name == "X" || name == "Y" || name == "Z" {
+            return Self::Pauli(name);
+        } else if name == "H" {
+            return Self::Hadamard(name);
+        } else if name == "PHASE" || name == "S" || name == "T" {
+            return Self::Phase(name);
+        } else if name == "CZ" || name == "CPHASE" {
+            return Self::ControlledPhase(name);
+        } else if name == "CNOT" || name == "CCNOT" {
+            return Self::ControlledX(name);
+        }
+
+        return Self::Unsupported(name);
+    }
```
This could/should be a `FromStr` impl, where an error is returned instead of the `Unsupported` variant

Shadow53 commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1530 @@
+    /// # Arguments
+    /// `name` - the name of the defined gate from instruction
+    /// `defgate` - a vector of all previously defined gates
+    fn get_supported_defgate(name: String, defgate: &Vec<String>) -> Self {
```
Return an error if the thing is not supported

Shadow53 commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1530 @@
+        // initialize circuit with empty wires of all qubits in program
+        let qubits = Program::get_used_qubits(&self);
+        for qubit in &qubits {
+            match qubit {
+                instruction::Qubit::Fixed(name) => {
+                    let wire = Wire {
+                        name: *name,
+                        ..Default::default()
+                    };
+                    diagram.circuit.insert(*name, Box::new(wire));
+                }
+                _ => (),
+            }
```
`if let`

Shadow53 commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1530 @@
+                                };
+
+                                // set parameters for phase gates
+                                if gate.name.contains("PHASE") {
```
Again, seems like something better handled with an enum than with strings

Shadow53 commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1530 @@
+            // set QW for any unused qubits in this instruction
+            diagram.set_qw(&qubits, &instruction);
+
+            match instruction {
+                // parse gate instructions into a new circuit
+                instruction::Instruction::Gate(gate) => {
+                    // for each qubit in a single gate instruction
+                    for qubit in &gate.qubits {
+                        match qubit {
+                            instruction::Qubit::Fixed(qubit) => {
```
Both of these `match`es should be `if let`s

Shadow53 commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1530 @@
+            body: body,
+            ..Default::default()
+        };
+        println!("{}", document.to_string());
```
This should be removed

Shadow53 reviewed (changes requested):
A lot of notes, largely about cleaning up code style. There are some larger design decisions I'd like to see changed as well. I'd also suggest running clippy locally, as it returns 41 warnings when I run it locally.

If you've been using `clippy`, remember that you need to pass either `--all-features` or `--features latex` for it to check feature-gated code that is not enabled by default.

`--all-targets` tells it to check tests as well.

```
cargo clippy --all-features --all-targets
```

Shadow53 reviewed (changes requested):
A lot of notes, largely about cleaning up code style. There are some larger design decisions I'd like to see changed as well. I'd also suggest running clippy locally, as it returns 41 warnings when I run it locally.

If you've been using `clippy`, remember that you need to pass either `--all-features` or `--features latex` for it to check feature-gated code that is not enabled by default.

`--all-targets` tells it to check tests as well.

```
cargo clippy --all-features --all-targets
```

hiyaryan commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1600 @@
+    fn default() -> Self {
+        Self {
+            header: r"\documentclass[convert={density=300,outext=.png}]{standalone}
+\usepackage[margin=1in]{geometry}
```
The header and footer are taken directly from pyQuil so I didn't make the decision myself.

notmgsk commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1600 @@
+    fn set_ctrl_targ(&mut self) -> Result<(), LatexGenError> {
+        // ensure every column preserves the connection between ctrl and targ
+        'column: for c in 0..=self.column {
+            let mut ctrls = vec![]; // the control qubits
```
@Shadow53 🤮

hiyaryan commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1291 @@
+            }
+        // gate is in canonical form, build wire
+        } else {
+            // set gates
```
I'm considering moving this block into a separate function called `set_gate`. However, I want to keep one of the arguments as `&mut self` but since this function, `parse_gate`, already has `&mut self`, I'm running into this compilation error: `cannot borrow `*self` as mutable because it is also borrowed as immutable`. My only way around this to set the function signature using mutable references from the fields I want to update in `Diagram` but that's not very elegant which is why I haven't moved this block yet.

hiyaryan replied:
Moved it to a method in `Wire` called `set_gate`. [Refactor gate setting from parse_gate into separate method set_gate.](https://github.com/rigetti/quil-rs/pull/145/commits/029f05b6c251b8e1195fa5e63212d6f563c19a2d)

notmgsk commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1291 @@
+    ///
+    /// # Arguments
+    /// `text` - a String representing a Greek or alphanumeric symbol
+    fn match_symbol(text: String) -> Symbol {
```
This should be a `impl From<String> for Symbol`, then wherever you use `Symbol::match_symbol(text)` would become `text.into()`.

notmgsk commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1291 @@
+    /// `&mut self` - exposes the wires on the Circuit
+    /// `qubits` - exposes the qubits used in the Program
+    /// `instruction` - exposes the qubits in a single Instruction
+    fn set_qw(&mut self, qubits: &HashSet<Qubit>, instruction: &Instruction) {
```
This name could be more descriptive

notmgsk commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1291 @@
+    UnsupportedGate { gate: String },
+}
+
+pub trait Latex {
```
IMO this should be `ToLatex` (c.f. `ToString`)

notmgsk replied:
Also I think it would be very funny if you named it `ToLaTeX`

notmgsk replied:
```
trait ToLatex {
    fn to_latex(self, settings: RenderSettings) -> Result<String, LatexGenError>;
}

impl<T> ToLatex for T {
    fn to_latex(self, settings: RenderSettings) -> Result<String, LatexGenError> {
        panic!("actually it's pronounced LaTeX")
    }
}
```

hiyaryan replied:
Renamed to `ToLatex`. It was previously named this. I think I liked the simplicity of `Latex`, but `ToLatex` is definitely more clear as to the purpose of this feature.

notmgsk reviewed (commented):
Another partial review :D

hiyaryan commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1291 @@
+
+impl Latex for Program {
+    /// Returns a Result containing a quil [`Program`] as a LaTeX string or a
+    /// [`LatexGenError`] defined using thiserror.
```
Removed, " defined using thiserror".

MarquessV commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1294 @@
+    /// `qubits` - exposes the qubits used in the Program
+    /// `instruction` - exposes the qubits in a single Instruction
+    fn set_empty(&mut self, qubits: &HashSet<Qubit>, instruction: &Instruction) {
+        'program_loop: for program_qubit in qubits {
+            if let Instruction::Gate(gate) = instruction {
+                for gate_qubit in &gate.qubits {
+                    if program_qubit == gate_qubit {
+                        continue 'program_loop;
+                    }
+                }
+
+                if let Qubit::Fixed(q) = program_qubit {
+                    if let Some(wire) = self.circuit.get_mut(q) {
+                        wire.empty.insert(self.column, Command::Qw);
+                    }
+                }
+            }
+        }
```
I think we can restructure this to only do something if `instruction` is a `Gate` right? YAlso, this was a fun one to play iterator golf on:

```suggestion
        if let Instruction::Gate(gate) = instruction {
            qubits
                .difference(&gate.qubits.iter().cloned().collect())
                .filter_map(|q| match q {
                    Qubit::Fixed(index) => Some(index),
                    _ => None,
                })
                .for_each(|index| {
                    self.circuit
                        .get_mut(index)
                        .map(|wire| wire.empty.insert(self.column, Command::Qw));
                });
        }

```

hiyaryan commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1294 @@
+    /// `CNOT` is `CONTROLLED X`
+    Cnot(String),
+    /// `CCNOT` is `CONTROLLED CONTROLLED X`
+    Ccnot(String),
```
According to the pyQuil docs it's a `CONTROLLED X`. That's what I got from it at least. Also, discussed this with Kalan in one of my status updates this week (I think from Monday).

hiyaryan replied:
It's the atomic form of a gate that's represented the same way on the diagram. You can think of CNOT as a shorthand notation for CONTROLLED X. Breaking the gate into this form allows the program to processes both Quil programs the same way.

MarquessV replied:
Yeah, we agree that `CNOT` == `CONTROLLED X` and are canonical gates. 

`CCNOT` is also a variant of this enum, but does it have an atomic circuit representation? Or is it a `CONTROLLED CNOT`?

hiyaryan replied:
So the atomic representation for CCNOT is a CONTROLLED CONTROLLED X. I get to this by breaking it first into CONTROLLED CNOT (intermediate step), then to CONTROLLED CONTROLLED X.

notmgsk commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1294 @@
+    /// # Arguments
+    /// `self` - exposes all attributes in the diagram
+    /// `gate` - the Gate of the Instruction from `to_latex`.
+    fn parse_gate(&mut self, gate: &Gate) -> Result<(), LatexGenError> {
```
Maybe more appropriately named `apply_gate`.

hiyaryan replied:
That name actually sounds a bit more consuming to me. I think because there's a `set_gate` already. Also, this method is doing analysis on the gate, that's why I chose the name `parse_gate`. How would you say `apply_gate` is a better name for this method?

hiyaryan replied:
Totally mixed up the words apply and insert after getting around to renaming this and changed the name to `insert_gate`. Maybe this name actually works better since I'm inserting the gate into the wire.

hiyaryan replied:
Updated to `apply_gate`!

hiyaryan commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,1294 @@
+        }
+
+        // get the first qubit in the BTreeMap
+        let first = circuit
+            .first_key_value()
+            .expect("previously checked that circuit is not empty")
+            .0
+            + 1;
```
```suggestion
        let first = circuit
            .first_key_value()
            .map(|wire| wire.0 + 1)
            .expect("previously checked that circuit is not empty");
```

Shadow53 commented on src/program/latex/mod.rs:
```diff
@@ -969,149 +795,83 @@ impl Latex for Program {
+                diagram.locate_targ()?;
+                diagram.column += 1;
+            } else if let Instruction::GateDefinition(_) = instruction {
+                // GateDefinition is supported and parsed in Gate
```
Does this mean that all `GateDefinition`s are turned into `Gate` elsewhere? Should indicate the function name/path for easier reference.

hiyaryan replied:
I'm not sure where this happens but somewhere when Quil is parsed in quil-rs `DEFGATE` is turned into `GATE`.

hiyaryan replied:
@Shadow53 I made this commit which maybe clarifies somethings a bit https://github.com/rigetti/quil-rs/commit/30169aa43747428105f9cd6e9ffb0220a514b0cd

hiyaryan commented:
This commit, [Organize code by restructuring design.](https://github.com/rigetti/quil-rs/pull/145/commits/c353c3fe8565ffec1813eb3d01fa2c985c5c0667), accomplishes the following,
- Removes `relationships` from `Diagram`.
- Removes `gate` argument from `set_gate`.
- Refactors `locate_targ` into two separate functions: `set_targ`, and `set_ctrl`.
- Moves `FoundTargetWithNoControl` to `to_latex` and is returned if the qubits in a gate are not unique.
- Renames `parse_gate` to `insert_gate`.
- Updates docs.

hiyaryan commented:
[Make ToLatex a method of Program.](https://github.com/rigetti/quil-rs/pull/145/commits/1ecc789b201788ff62f2ad96d10b780c22d237be) Downgrades `ToLatex` from trait to method `to_latex` of `Program` implemented in mod `latex` since this feature is only implemented for `Program`.

[Add derive_more, change From<String> to FromStr, generify with From<S\>.](https://github.com/rigetti/quil-rs/pull/145/commits/4e4ca5a339bbfd6cadf1576bc0a7156c9022a1e0) Makes conversions from enum to String and String to enum more concise, generic, and optimizes code preventing unnecessary cloning.

hiyaryan commented on src/program/latex/mod.rs:
```diff
@@ -58,9 +58,9 @@ enum Command {
-        match command {
+impl ToString for Command {
+    fn to_string(&self) -> String {
+        match self {
```
`Command` looks different now but I'll keep this in mind!

hiyaryan commented on src/program/latex/mod.rs:
```diff
@@ -96,47 +99,63 @@ enum Symbol {
+
+impl From<String> for Symbol {
+    fn from(text: String) -> Self {
+        match text.as_str() {
```
Thank you! Made those changes here [Add derive_more, change From<String> to FromStr, generify with From<S\>.](https://github.com/rigetti/quil-rs/pull/145/commits/4e4ca5a339bbfd6cadf1576bc0a7156c9022a1e0)

hiyaryan commented on src/program/latex/mod.rs:
```diff
@@ -171,70 +190,59 @@ impl Default for Settings {
+        let last = circuit
+            .last_key_value()
+            .map(|wire| wire.0 - 1)
+            .expect("previously checked that circuit has at least two wires");
```
I like it! Updated current code to this but needed to change `keys_iter()` to `keys_iter` on lines 214 and 219.

hiyaryan commented on src/program/latex/mod.rs:
```diff
@@ -278,294 +286,236 @@ impl Display for Document {
+            if let Some(last) = relationship.last() {
+                if let Qubit::Fixed(targ) = last {
+                    // any qubit before last in the relationship
+                    let mut pred = None;
```
This method has been removed.

hiyaryan commented on src/program/latex/mod.rs:
```diff
@@ -580,119 +530,110 @@ impl Display for Diagram {
 
             // if this is the last key iteration, omit Nr from end of line
             if i < self.circuit.len() - 1 {
                 // indicate a new row
```
When the circuit is formatted to LaTeX is goes row by row. The outer loop takes each row, and then there's an inner loop that writes out the columns for that row. Once it breaks from the inner loop, which is where this part of the code is, the next wire is ready to be serialized on a new line, this is indicated using the `Nr` command which is the equivalent of `\n` in LaTeX.

hiyaryan commented on src/program/latex/mod.rs:
```diff
@@ -580,119 +530,110 @@ impl Display for Diagram {
+                        // concatenate superscripts
+                        if !superscript.is_empty() {
+                            _gate.push_str(&superscript);
                         }
```
Yes, didn't need it.

hiyaryan commented on src/program/latex/mod.rs:
```diff
@@ -580,119 +530,110 @@ impl Display for Diagram {
+                                    )
+                                }
                             }
                         }
```
If `is_none` for `ctrl` or `targ` then the superscript can still be added to single gates. This is used outside of these conditionals.

hiyaryan commented:
[Apply suggestions to extend derived traits and shorten/simplify code.](https://github.com/rigetti/quil-rs/pull/145/commits/46c024f2b21ca0176a0e26e9f9ccd5eccca90315) Addresses most of @Shadow53 reviews. (Please see discussion in comments for unresolved conversations). As stated in commit message, this commit makes minor changes shortening an simplifying code in `insert_gate`, `impute_missing_qubits`, and `Diagram::fmt`. It also includes derived traits, `PartialEq`, `Eq`, and `Hash`, for all enums.

hiyaryan commented:
Previous commits update the docs and organizes the latex module and struct methods. The following is now the module structure of latex,

-- latex/
---- diagram/
------ mod.rs
------ wire.rs
---- snapshots/
---- mod.rs
---- settings.rs

hiyaryan commented:
Previous commits refactors code removing duplication, making it easier to read, makes it more concise by taking advantage of built-in Rust functions for vectors, and improves performance by removing type conversion where possible. These changes can be seen most prevalently in Diagram's `Display::fmt` and `Wire` struct. 

Further, to make adding superscripts to phase and generic gates more generalized, a new String inner type was added to their respective RenderCommand variants.

Additionally, `DAGGER`, while stated to be implemented, was found do be missing from `PHASE`/`CPHASE`. This was implemented and new snapshots were added. The following instructions demo the new implementation.

`DAGGER CPHASE(alpha) 0 1`
```latex
\documentclass[convert={density=300,outext=.png}]{standalone}
\usepackage[margin=1in]{geometry}
\usepackage{tikz}
\usetikzlibrary{quantikz}
\begin{document}
\begin{tikzcd}
\lstick{\ket{q_{0}}} & \ctrl{1} & \qw \\
\lstick{\ket{q_{1}}} & \phase{\alpha^{\dagger}} & \qw
\end{tikzcd}
\end{document}
```
<img width="556" alt="Screenshot 2023-03-31 at 4 34 26 PM" src="https://user-images.githubusercontent.com/58452495/229250298-4d8675e6-5001-4339-9374-f942f19ea78e.png">

`DAGGER PHASE(pi) 0`
```latex
\documentclass[convert={density=300,outext=.png}]{standalone}
\usepackage[margin=1in]{geometry}
\usepackage{tikz}
\usetikzlibrary{quantikz}
\begin{document}
\begin{tikzcd}
\lstick{\ket{q_{0}}} & \phase{\pi^{\dagger}} & \qw
\end{tikzcd}
\end{document}
```
<img width="558" alt="Screenshot 2023-03-31 at 4 36 05 PM" src="https://user-images.githubusercontent.com/58452495/229250305-835cf292-499d-4d24-8462-909ab006ec0e.png">

jselig-rigetti commented on src/program/latex/diagram/mod.rs:
```diff
@@ -0,0 +1,174 @@
+            write!(f, "{}", &RenderCommand::Qw)?;
+
+            // if this is the last iteration, omit a new row from the end of the line
+            if *qubit != *self.circuit.keys().last().unwrap() {
```
is it worth using `for (i, (qubit, wire)) in &self.circuit.enumerate() {` or similar to make the check here simpler? Or cache `let last = self.circuit.keys().last().unwrap()` before the loop? Second one might need extra handling if `self.circuit.len() == 0`.

Not a blocker, just curious other's thoughts

hiyaryan replied:
Similar to the last previous review I was able to pull it outside of the for-loop in general. I was able to handle the len==0 case with, `let last = self.circuit.keys().last().unwrap_or(&0);` Now it only needs to run once.

hiyaryan commented on src/program/latex/diagram/mod.rs:
```diff
@@ -0,0 +1,174 @@
+                if let Some(wire) = self.circuit.get_mut(instruction_qubit) {
+                    // set the control and target qubits
+                    if gate.qubits.len() > 1 || canonical_gate == "PHASE" {
+                        // set the target qubit if the qubit is equal to the last qubit in gate
+                        if qubit == gate.qubits.last().unwrap() {
+                            wire.set_targ();
+                        // otherwise, set the control qubit
+                        } else {
+                            wire.set_ctrl(qubit, gate.qubits.last().unwrap(), &circuit_qubits);
+                        }
```
Looks like I can even pull it outside of the for-loop and call it once.

kalzoo commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,695 @@
+/// inside `backticks`.
+#[derive(Clone, Debug, derive_more::Display, PartialEq, Eq, Hash)]
+pub(crate) enum RenderCommand {
+    /// Make a qubit "stick out" from the left.
+    #[display(fmt = "\\lstick{{\\ket{{q_{{{_0}}}}}}}")]
+    Lstick(u64),
```
Design note: if the principle of `RenderCommand` is that it should faithfully and literally represent the underlying `Quantikz` syntax, then `LStick(u64)` diverges from that, because `\lstick{}` can contain any other Quantikz elements.

This approach is ok too - as a rendering helper - but I would suggest making it more intuitive in that case. ie, `Lstick` -> `LeftWireLabel`. 

`Lstick(u64)` straddles two layers of abstraction.

hiyaryan replied:
Updated to `LeftWireLabel` [Change Lstick to LeftWireLabel.](https://github.com/rigetti/quil-rs/pull/145/commits/11191e08839aff4430a916dada22b4743e06b648)

hiyaryan commented on src/program/latex/diagram/mod.rs:
```diff
@@ -0,0 +1,174 @@
+    /// `CZ` is `CONTROLLED Z`
+    #[display(fmt = "Z")]
+    Cz,
+}
```
@kalzoo This commit [Use regex for canonical gate decomposition and handle unimplemented X…](https://github.com/rigetti/quil-rs/pull/145/commits/68cc16eeec162a13ac3246a8cfa5fd1978c78546) removes `CanonicalGate` as an enum, and now decomposes gates into canonical form in `apply_gate`. It also handles the unimplemented `XY` gate. Since `XY` is a built-in gate, I think the best way to handle it is to look for it explicitly since any other gate of the same form could be a `DEFGATE`. So I am handling this in the same block where I look for multi-qubit gates, and explicitly checking if the canonical gate name is XY. If it is, it returns an `UnsupportedGate` error.

hiyaryan replied:
Using `CONTROLLED` count to determine if a gate is a control/target gate. See [Use the generic gate and counted modifiers, and catch unsupported gat…](https://github.com/rigetti/quil-rs/pull/145/commits/755883df424663d2c49a222c9d2d90b3abe73c26)

hiyaryan commented on src/program/latex/diagram/wire.rs:
```diff
@@ -0,0 +1,202 @@
+    /// at this column the wire is a control some distance from the target
+    pub(crate) ctrl: HashMap<usize, i64>,
+    /// at this column is the wire a target?
+    pub(crate) targ: HashMap<usize, bool>,
```
Updated to a single gate vector of generic gate types. See [Use the generic gate and counted modifiers, and catch unsupported gat…](https://github.com/rigetti/quil-rs/pull/145/commits/755883df424663d2c49a222c9d2d90b3abe73c26)

hiyaryan commented on src/program/latex/diagram/mod.rs:
```diff
@@ -0,0 +1,174 @@
+pub(super) struct Diagram {
+    /// customizes how the diagram renders the circuit
+    pub(crate) settings: RenderSettings,
+    /// a BTreeMap of wires with the name of the wire as the key
```
By qubit index do you mean qubit name? I'm using the u64 of a Fixed Qubit as the key.

hiyaryan commented on src/program/latex/diagram/mod.rs:
```diff
@@ -0,0 +1,174 @@
+            write!(f, "{wire}")?;
+
+            // chain an empty column to the end of the line
+            write!(f, " & ")?;
+            write!(f, "{}", &RenderCommand::Qw)?;
```
The way I'm handling symbol is by finding what parameters are contained in a gate instruction then matching it with it's symbol enum variant which contains a reference to it's TeXified version or simply, textual, version (depending on the settings or if its implemented). This symbol is stored in parameter, then when I'm ready to use it I simply to_string it. It seems like I am handling it the same way?

hiyaryan commented on src/program/latex/diagram/mod.rs:
```diff
@@ -0,0 +1,176 @@
+            .unwrap_or(gate.name.clone());
+
+        // get the names of the qubits in the circuit before circuit is borrowed as mutable
+        let circuit_qubits: Vec<u64> = self.circuit.keys().cloned().collect();
```
I can see how this works with qubits that are imputed but not the other way. Lets say we have a range of 2 for qubits [2,5] and 5 controls 2, so it should be 5{-1} (5 is a distance of -1 from 2). So if I use the range to find this distance, then 5-2 = 3, and 2-2 = 0, 0-3 = -3. Even if I make the adjustment to range -1, so 1, then I still end up with -3, 5-1 = 4, 2-1 = 1, 1-4 = -3. 

I get that we can use the range to shift the qubits, but this works well when the qubits are imputed. Otherwise it doesn't shift the qubits correctly as it assumes they're imputed. I'm not seeing a very straight forward way to use this. I find is easier to locate these qubits in the full circuit by index and just subtract the index. How are you using this in your version?

kalzoo commented on src/program/latex/diagram/mod.rs:
```diff
@@ -0,0 +1,151 @@
+            if let Qubit::Fixed(instruction_qubit) = qubit {
+                if let Some(wire) = self.circuit.get_mut(instruction_qubit) {
+                    // push the gate to the wire
+                    wire.gates.push(wire::T::try_from(gate.clone())?);
+
+                    // update the last gate to a control gate if the instruction has more than one qubit
+                    if gate.qubits.len() > 1 {
+                        if let wire::T::StdGate {
+                            name,
+                            dagger_count: _,
+                            ctrl_count,
+                        } = wire.gates.last().unwrap()
```
Here, you push a gate onto the `Vec` and then pull it off and `unwrap()`. Why not just create it and use it directly?

Also, you discard information when you `T::try_from(gate.clone())`. You have a `gate`, which you know is a Gate - then you have a `T`, which might be a gate or something else. The Gate should be converted into something that is _more specific_ than a gate, not less. As from our last 1x1,  there should be a `struct` with the five fields derived from that gate. Then you can operate on that newer, more specific object directly.

MarquessV commented on src/program/latex/diagram/wire.rs:
```diff
@@ -0,0 +1,198 @@
+
+use super::super::{LatexGenError, Parameter, RenderCommand, Symbol};
+
+/// The QuantikzGate is a valid cell of a QuantikzCellType that is the form of
+/// the ``Gate`` struct in the ``Instruction`` containing the components that
+/// are used to render a gate on the wire in the circuit diagram.
```
This is a bit wordy, and probably shouldn't reference `QuantizCellType`, `Gate`, or `Instruction`. It stands on it's own as a type. Did AI write this per chance?

hiyaryan replied:
It was AI assisted but I went through and reworded it pretty much entirely. If it stands on its own then I could just remove the doc.

MarquessV commented on src/program/latex/diagram/wire.rs:
```diff
@@ -0,0 +1,198 @@
+pub(crate) struct Wire {
+    /// the columns on the wire that can be serialized into LaTeX
+    pub(crate) columns: Vec<QuantikzColumn>,
+}
```
I feel like there is a conflict with the naming here. The name `Wire` (and docstring) implies it should represent a single row/qubit but it actually contains a vector of columns.

I think "column" is the right abstraction here, that is, each instruction in the program corresponds to a single column in the diagram. The docstring hints at this kind of structure, but it's not modeled by the data structures being used.

Sometimes it's helpful to start with the data structures, then build an implementation on top of that. What if we started with these:

```rust
struct Diagram {
    settings: RenderSettings,
    wires: HashMap<usize, usize>, // qubit index -> column index
    columns: Vec<DiagramColumn>
}

struct DiagramColumn {
    cells: Vec<QuantikzCell>
}

struct QuantikzCellType {
    Gate(QuantikzGate),
    Control(i64),
    Target,
    Empty,
}

struct QuantikGate {
    name: String,
    parameters: Vec<Parameter>,
    dagger_count: usize,
}
```

At a glance, all of these names feel a bit more clear to me. We have a diagram, which is composed of columns, and each column is composed of individual cells with a quantikz command. We also store the mapping of the qubit indices to their corresponding column index.

An API build on top of this might look something like:

```rust
impl Diagram {
    fn new(qubits: Vec<int>, settings: RenderSettings) -> Self {
         // Initialize a new `Diagram` with the given qubits
         // If settings.impute_missing qubits, then pad the gaps.
    }

    fn add_gate(&mut self, gate: &Gate) -> Result<(), LatexGenError> {
        // Initialize a column full of empty cells to be filled in.
        let mut column: Vec<QuantikzColumn> = vec![QuantikzCellType::Empty; len(self.wires)] 
        
        // This is oversimplified for the example. There will be a little
        // bit of pre-processing necessary to get your "canonical gate",
        // and it probably makes sense to go in reverse order so you can start
        // with the target qubit.
        for each qubit in gate.qubits.map(|q| q.into_fixed())? {
            column_index = self.wires.get(qubit)
            // set column[column_index] to the correct QuantikzCellType
        }

        self.columns.push(column)
    }
}
```

Does this make sense as an outline? Is it clear how `to_latex` would interact with `Diagram`?

hiyaryan commented on Cargo.toml:
```diff
@@ -1,17 +1,20 @@
 [dependencies]
+derive_more = { version = "0.99.17", optional = true }
 dot-writer = { version = "0.1.2", optional = true }
+enum-as-inner = { version = "0.5.1", optional = true }
```
That makes sense. Originally this dependency was simply `enum-as-inner = "0.5.1"`, so I'll revert the commit I made that changed it to this. I had been making everything I imported for `to_latex` as optional.

hiyaryan commented on src/program/latex/mod.rs:
```diff
@@ -0,0 +1,748 @@
+
+        for instruction in instructions {
+            // parse gate instructions into a new circuit
+            if let Instruction::Gate(gate) = instruction {
```
I think before it made sense to use `if let` because I was working with only the Gate instruction but it makes sense to refactor it to `match` now.

Shadow53 commented on src/program/latex/diagram/mod.rs:
```diff
@@ -0,0 +1,176 @@
+
+        let mut qubit_iterator = gate
+            .qubits
+            .iter()
+            .map(|q| {
+                q.clone()
```
Can also do

```suggestion
            .iter()
            .cloned()
            .map(|q| {
                q
```

Shadow53 commented on src/program/latex/diagram/mod.rs:
```diff
@@ -0,0 +1,176 @@
+        cells[target_index] = QuantikzCell::Gate(quantikz_gate);
+
+        for qubit in qubit_iterator {
+            let column_index = *self
+                .wires
+                .get(&qubit?)
```
If I understand right what's happening here, I think this reads more clearly:

```suggestion
            let column_index = self
                .wires
                .get(&qubit?)
                .copied()
```

Shadow53 commented on src/program/latex/diagram/mod.rs:
```diff
@@ -0,0 +1,176 @@
+
+impl fmt::Display for Diagram {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let last = self.wires.keys().last().unwrap_or(&0);
```
```suggestion
        let last = self.wires.keys().last().copied().unwrap_or(0);
```

Can also replace `unwrap_or(0)` with `unwrap_or_default`

Shadow53 commented on src/program/latex/diagram/mod.rs:
```diff
@@ -0,0 +1,176 @@
+                    f,
+                    "{}{}",
+                    &RenderCommand::Separate,
+                    command.to_latex(&cell_settings).map_err(|_| fmt::Error)?
```
I'm not a fan of erasing errors like this. See my note above about having a `display_latex` function, which avoids this.

From the docs for `std::fmt::Error`:

> This type does not support transmission of an error other than that an error occurred. Any extra information must be arranged to be transmitted through some other means.

IMO, the only error that should come from `Display` is to do with the actual writing of text, not validating the thing to write

Shadow53 reviewed (changes requested):
This is looking really good and I like how things have simplified and been split into modules with their own concerns.

Some high level suggestions, some of which are repeats of comments I left:

First, make sure you run `cargo clippy --all-features` -- there are some things like `format!("{}", value)` that would be caught by it.

Second, there's an odd mix of `to_latex` and `fmt::Display`, which leads to things like erasing errors that may be useful to the end user. My suggestion for this is to follow a pattern I've seen in a few places, where you have the "real" `struct`, e.g. `QuantikzCell`, and a function that creates a *new* struct, e.g. `QuantikzCellDisplay<'a>`, which is *known* to be in a good state and implements `fmt::Display`. This:

- Helps decrease the number of owned strings being passed around, which should optimize the runtime a little
- Separates the validation of the input from printing the output, so there's no more replacing a useful error with `fmt::Error`
- Can be combined to build larger structs that also implement `fmt::Display`, e.g.
   ```rust
   struct ProgramDisplay<'a> {
       cells: Vec<QuantikzCellDisplay<'a>>,
       // etc.
   }
   ```

hiyaryan commented:
>  First, make sure you run cargo clippy --all-features -- there are some things like format!("{}", value) that would be caught by it.

That's strange, doesn't look like clippy catches the issues you mentioned @Shadow53. I have clippy --fix set automatically whenever I save. I also ran `cargo clippy --all-features --fix --allow-dirty` just to check and still, the same.

Shadow53 commented:
> > First, make sure you run cargo clippy --all-features -- there are some things like format!("{}", value) that would be caught by it.
> 
> That's strange, doesn't look like clippy catches the issues you mentioned @Shadow53. I have clippy --fix set automatically whenever I save. I also ran `cargo clippy --all-features --fix --allow-dirty` just to check and still, the same.

Odd. Might be something that I'm used to having enabled in my projects that isn't enabled by default or in `src/lib.rs`

hiyaryan commented:
> Second, there's an odd mix of `to_latex` and `fmt::Display`, which leads to things like erasing errors that may be useful to the end user. My suggestion for this is to follow a pattern I've seen in a few places, where you have the "real" `struct`, e.g. `QuantikzCell`, and a function that creates a _new_ struct, e.g. `QuantikzCellDisplay<'a>`, which is _known_ to be in a good state and implements `fmt::Display`. This:

@Shadow53 I haven't heard from anyone on the team on the feedback you requested from them on this. Is this something worth pursuing on this PR?
